src/main/scala/com/cognite/spark/datasource/DataPointsRelationV1.scala

```diff
@@ -0,0 +1,299 @@
+    val indicesOfRequiredFields = requiredColumns.map(f => fieldNamesInOrder.indexOf[String](f))
+    val rowOfAllFields = itemToRow(item)
+    Row.fromSeq(indicesOfRequiredFields.map(idx => rowOfAllFields.get(idx)))
+  }
+
+  def itemToRow(item: DataPointsItem): Row =
+    Row(
+      item.name,
+      item.id,
+      item.externalId,
+      item.timestamp,
+      item.value,
+      item.aggregation,
+      item.granularity
+    )
+
+  override def buildScan(requiredColumns: Array[String], filters: Array[Filter]): RDD[Row] =
+    DataPointsRddV1(sqlContext.sparkContext, config, getIOs(filters), toRow(requiredColumns))
+  def getIOs(filters: Array[Filter])(
+      client: GenericClient[IO, Nothing]): Seq[(DataPointsFilter, IO[Seq[SdkDataPoint]])] = {
+    val pushdownFilterExpression = toPushdownFilterExpression(filters)
+    val filtersAsMaps = pushdownToParameters(pushdownFilterExpression)
+
+    val ids = filtersAsMaps.flatMap(m => m.get("id")).map(_.toLong)
+    val externalIds = filtersAsMaps.flatMap(m => m.get("externalId"))
+
+    val (lowerTimeLimit, upperTimeLimit) = filtersToTimestampLimits(filters)
+    val (aggregations, granularities) = getAggregationSettings(filters)
+
+    if (aggregations.isEmpty) {
+      ids.map(
+        id =>
+          (
+            DataPointsFilter(Some(id), None, None, None),
+            client.dataPoints.queryById(id, lowerTimeLimit, upperTimeLimit)))
+    } else {
+      getIOsWithAggregates(
+        ids,
+        externalIds,
+        lowerTimeLimit,
+        upperTimeLimit,
+        granularities,
+        aggregations,
+        client)
+    }
+  }
+
+  def getIOsWithAggregates(
+      ids: Seq[Long],
+      externalIds: Seq[String],
+      lowerTimeLimit: Long,
+      upperTimeLimit: Long,
+      granularities: Seq[String],
+      aggregations: Seq[AggregationFilter],
+      client: GenericClient[IO, Nothing]): Seq[(DataPointsFilter, IO[Seq[SdkDataPoint]])] = {
+    val ioFromId = for {
+      id <- ids
+      granularity <- granularities
+    } yield {
+      (
+        DataPointsFilter(Some(id), None, Some(aggregations.map(_.aggregation)), Some(granularity)),
+        client.dataPoints
+          .queryAggregatesById(
+            id,
+            lowerTimeLimit,
+            upperTimeLimit,
+            granularity,
+            aggregations.map(_.aggregation))
+          .map(m => m.toSeq.flatMap(t => t._2)))
+    }
+
+    val ioFromExternalId = for {
+      extId <- externalIds
+      granularity <- granularities
+    } yield {
+      (
+        DataPointsFilter(
+          None,
+          Some(extId),
+          Some(aggregations.map(_.aggregation)),
+          Some(granularity)),
+        client.dataPoints
+          .queryAggregatesByExternalId(
+            extId,
+            lowerTimeLimit,
+            upperTimeLimit,
+            granularity,
+            aggregations.map(_.aggregation))
+          .map(m => m.toSeq.flatMap(t => t._2)))
+    }
+    ioFromId ++ ioFromExternalId
```

Comment:
Use `merge` to make a concurrent stream, otherwise it will start fetching from `ioFromExternalId` only after `ioFromId` has been emptied.

Reply:
This merges two `Seq[(DataPointsFilter, IO[stuff])` that are picked by index in compute, similar to the way `EventsRelationV1` uses a `Seq[Stream]`. Should be ok this way?

Reply:
Ah, my bad, I misunderstood the usage.